arlo_server/ballot_manifest.py

```diff
@@ -0,0 +1,157 @@
+import csv, io, locale, uuid
+from sqlalchemy.orm.session import Session
+from flask import request, jsonify, Request
+from werkzeug.exceptions import BadRequest
+from datetime import datetime
+
+from arlo_server import app
+from arlo_server.models import (
+    db,
+    Batch,
+    Election,
+    File,
+    Jurisdiction,
+)
+from arlo_server.auth import with_jurisdiction_access
+from util.process_file import (
+    process_file,
+    serialize_file,
+    serialize_file_processing,
+    UserError,
+)
+
+BATCH_NAME = "Batch Name"
+NUMBER_OF_BALLOTS = "Number of Ballots"
+STORAGE_LOCATION = "Storage Location"
+TABULATOR = "Tabulator"
+
+
+def process_ballot_manifest_file(
+    session: Session, jurisdiction: Jurisdiction, file: File
+):
+    assert jurisdiction.manifest_file_id == file.id
+
+    def process():
+        manifest_csv = csv.DictReader(io.StringIO(file.contents))
+
+        missing_fields = [
+            field
+            for field in [BATCH_NAME, NUMBER_OF_BALLOTS]
+            if field not in manifest_csv.fieldnames
+        ]
+
+        if missing_fields:
+            raise UserError(f"Missing required CSV fields: {', '.join(missing_fields)}")
+
+        num_batches = 0
+        num_ballots = 0
+        for row in manifest_csv:
+            num_ballots_in_batch_csv = row[NUMBER_OF_BALLOTS]
+
+            try:
+                num_ballots_in_batch = locale.atoi(num_ballots_in_batch_csv)
+            except ValueError as error:
+                raise UserError(
+                    f"Invalid value for '{NUMBER_OF_BALLOTS}' on line {manifest_csv.line_num}: {num_ballots_in_batch_csv}"
+                ) from error
+
+            batch = Batch(
+                id=str(uuid.uuid4()),
+                name=row[BATCH_NAME],
+                jurisdiction_id=jurisdiction.id,
+                num_ballots=num_ballots_in_batch,
+                storage_location=row.get(STORAGE_LOCATION) or None,
+                tabulator=row.get(TABULATOR) or None,
+            )
+            session.add(batch)
+            num_batches += 1
+            num_ballots += batch.num_ballots
+
+        jurisdiction.manifest_num_ballots = num_ballots
+        jurisdiction.manifest_num_batches = num_batches
+
+    process_file(session, file, process)
+
+    election = jurisdiction.election
+
+    # If we're in the single-jurisdiction flow, posting the ballot manifest
+    # starts the first round, so we need to sample the ballots.
+    # In the multi-jurisdiction flow, this happens after all jurisdictions
+    # upload manifests, and is triggered by a different endpoint.
+    if not election.is_multi_jurisdiction:
+        # Import this here to avoid circular dependencies
+        from arlo_server.routes import sample_ballots
```

Comment:
I moved this import here to accommodate the division of logic between the two files. `routes` has all the logic that is for the single-page flow generally, but we need to include `sample_ballots` for this one case here. This file has some logic that's shared for saving/deleting ballot manifests, so it needs to be imported in `routes`.